python/ray/experimental/workflow/workflow_storage.py

```diff
@@ -0,0 +1,268 @@
+"""
+This module is higher-level abstraction of storage directly used by
+workflows.
+"""
+
+import json
+import pathlib
+from typing import Dict, List, Optional, Any, Callable, Tuple, Union
+
+from dataclasses import dataclass
+
+import ray
+from ray.experimental.workflow import storage
+from ray.experimental.workflow.common import Workflow, WorkflowInputs, StepID
+from ray.experimental.workflow import workflow_context
+from ray.experimental.workflow import serialization_context
+
+from ray.experimental.workflow.constants import (
+    STEPS_DIR, OBJECTS_DIR, STEP_INPUTS_METADATA, STEP_OUTPUTS_METADATA,
+    STEP_FUNC_BODY, STEP_ARGS, STEP_OUTPUT, STEP_OUTPUTS_FORWARD)
+
+
+@dataclass
+class StepInspectResult:
+    args_valid: bool = False
+    func_body_valid: bool = False
+    object_refs: Optional[List[str]] = None
+    workflows: Optional[List[str]] = None
+    output_object_valid: bool = False
+    output_step_id: Optional[StepID] = None
+
+    def is_recoverable(self) -> bool:
+        return (self.args_valid and self.object_refs is not None
+                and self.workflows is not None and self.func_body_valid)
+
+
+class WorkflowStorage:
+    """Access workflow in storage. This is a high-level function,
+    which does not care about the underlining storage implementation."""
+
+    def __init__(self,
+                 workflow_id: Optional[str] = None,
+                 store: Optional[storage.Storage] = None):
+        if workflow_id is None:
+            context = workflow_context.get_workflow_step_context()
+            workflow_id = context.workflow_id
+        if store is None:
+            store = storage.get_global_storage()
+        self._storage = store
+        self._workflow_dir = pathlib.Path(workflow_id)
+        self._steps_dir = self._workflow_dir / STEPS_DIR
+        self._objects_dir = self._workflow_dir / OBJECTS_DIR
+
+    def write_step_inputs(self, step_id: StepID, inputs: WorkflowInputs):
+        """Save workflow inputs."""
+        args = inputs.args
+        f = inputs.func_body
+        metadata = {
+            "step_id": inputs.step_id,
+            "name": f.__module__ + "." + f.__qualname__,
+            "object_refs": inputs.object_refs,
+            "workflows": inputs.workflows,
+        }
+        with serialization_context.workflow_args_keeping_context():
+            # TODO(suquark): in the future we should write to storage directly
+            # with plasma store object in memory.
+            args_obj = ray.get(args)
+        step_dir = self._steps_dir / step_id
+        if not self._storage.directory_exists(step_dir):
+            self._storage.create_directory(step_dir)
+        self._storage.write_json_atomic(metadata,
+                                        step_dir / STEP_INPUTS_METADATA)
+        self._storage.write_object_atomic(f, step_dir / STEP_FUNC_BODY)
+        self._storage.write_object_atomic(args_obj, step_dir / STEP_ARGS)
```

Comment:
Could we write these into a staging dir and then do an atomic move of the directory to the final location? Otherwise seems like we could end up in a bad state if we crash in the middle.